presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java

```diff
@@ -593,6 +597,24 @@ public PlanOptimizers(
         // We run it again to mark this for intermediate join nodes.
         builder.add(new StatsRecordingPlanOptimizer(optimizerStats, new HistoricalStatisticsEquivalentPlanMarkingOptimizer(statsCalculator)));
 
+        // Run this set of join transformations after ReorderJoins, but before DetermineJoinDistributionType
+        builder.add(new IterativeOptimizer(
+                ruleStats,
+                statsCalculator,
+                estimatedExchangesCostCalculator,
+                Optional.of(new LogicalPropertiesProviderImpl(new FunctionResolution(metadata.getFunctionAndTypeManager()))),
+                ImmutableSet.of(
+                        new TransformDistinctInnerJoinToRightEarlyOutJoin(),
+                        new TransformDistinctInnerJoinToLeftEarlyOutJoin(),
```

Comment:
If I understand correctly, enabling TransformUncorrelatedInPredicateSubqueryToDistinctInnerJoin will effectively disable TransformUncorrelatedInPredicateSubqueryToSemiJoin. 
And these two rewrite rules here are restoring to the current behavior if the join is not reordered. Is this correct?